### PR TITLE
Fix integer division semantics in the arithmetic solver

### PR DESCRIPTION
**Pull request by benjaminfjones**
*Description of changes:*

Integer division (by a constant) was being treated the same as real division. This change fixes the bug by introducing new constraints based on Euclidean division semantics. 

New unit tests are provided that cover both upper and lower remainder bounds in the semantics.

----
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/arithmetic/lia/solver_result_api.rs [hunk 1]
```diff
@@ -31,17 +31,18 @@ impl SolverDecisionApi {
...
+                        var_term_map.get(&var).map(|term| (term.clone(), value))
+                    })
                     .collect();
                 Ok(SolverDecisionApi::FEASIBLE(Assignment::new(term_assign)))
             }
             SolverDecision::INFEASIBLE(conflict) => {
                 let term_conflict: Conflict<Term> = conflict
                     .iter()
-                    .map(|v| var_term_map.get(v).unwrap().clone())
+                    .filter_map(|v| var_term_map.get(v).cloned())
```
Comment (amarshah1) [inline src/arithmetic/lia/solver_result_api.rs:45]: why do we need a filter_map instead of a map here?

File: src/arithmetic/lia/frontend.rs [hunk 2]
```diff
@@ -1067,10 +1111,99 @@ mod tests {
...
+    /// (not 0, which exact division toward zero would give, and not -0.5 which
+    /// the pre-fix real-division handling produced).
+    #[test]
+    fn test_int_div_negative_numerator_floors() {
+        let smt_input = r#"
+(set-logic QF_LIA)
+(declare-const r Int)
+(assert (= r (div (- 1) 2)))  ; Euclidean: -1 = 2*(-1) + 1, so r = -1
+(assert (= r (- 1)))
+(check-sat)
```
Comment (amarshah1) [inline src/arithmetic/lia/frontend.rs:1221]: can you add one more regression test for when we have a negative divisor?

File: src/arithmetic/lia/frontend.rs [hunk 3]
```diff
@@ -239,6 +239,60 @@ fn convert_linear_term(ctx: &mut ConvContext, term: &Term) -> FrontendResult<Lin
...
                     }
                 }
+                INT_DIV_SYMBOL => {
+                    // SMT-LIB integer division (Euclidean): (div e n) = q
+                    // where e = n * q + r, 0 <= r < |n|
+                    debug_assert!(args.len() == 2, "chainable division is not supported");
+                    let e1 = convert_linear_term(ctx, &args[0])?;
+                    let e2 = convert_linear_term(ctx, &args[1])?;
+                    if let Some(n) = e2.is_constant() {
+                        if n.is_zero() {
```
Comment (amarshah1) [inline src/arithmetic/lia/frontend.rs:249]: I think SMT-Lib defines division by zero as an uninterpreted function. At least thats how Z3 behaves. Not sure how we want to tackle these in Sundance

**Comment by amarshah1** [comment]
This addresses some of the issues described in #117 , but not everything because there are some fixes that need to be made outside of the arithmetic solver. But pretty good overall!

**Comment by amarshah1** [comment]
Do we have a similar issue with Integer Mod?

**Comment by amarshah1** [comment]
> Do we have a similar issue with Integer Mod?

It seems like the internal solver doesnt handle mod at all? I think that is out of scope of this PR, but we should capture this as an issue

File: src/arithmetic/lia/frontend.rs (hunk 3 above)
Comment (amarshah1) [reply src/arithmetic/lia/frontend.rs:249]: I don't think this needs to be addressed in this PR necessarily if its too complicated, but it should be captured as an issue

**Review by amarshah1 — APPROVED** [review]
Looks good overall. Can you capture the divide by 0 and the integer mod things as two separate issues if you think they are problems that we need to address and are out of scope for this PR?

**Comment by benjaminfjones** [comment]
> > Do we have a similar issue with Integer Mod?
> 
> It seems like the internal solver doesnt handle mod at all? I think that is out of scope of this PR, but we should capture this as an issue

Yeah, the arithmetic solver doesn't handle `mod` at all. I'll make a sub-issue.

File: src/arithmetic/lia/frontend.rs (hunk 3 above)
Comment (benjaminfjones) [reply src/arithmetic/lia/frontend.rs:249]: I'll create an issue; it's definitely a gap

File: src/arithmetic/lia/solver_result_api.rs (hunk 1 above)
Comment (benjaminfjones) [reply src/arithmetic/lia/solver_result_api.rs:45]: Sorry, this was unrelated to the change at hand. I thought it would be cleaner to `filter_map` the `get` rather than map get and then unwrap everything. It's an invariant that every var in the conflict is in the `var_term_map` so the result is always the same either way.

File: src/arithmetic/lia/frontend.rs (hunk 3 above)
Comment (benjaminfjones) [reply src/arithmetic/lia/frontend.rs:249]: https://github.com/yaspar-org/Sundance-SMT/issues/133

**Comment by benjaminfjones** [comment]
> > > Do we have a similar issue with Integer Mod?
> > 
> > 
> > It seems like the internal solver doesnt handle mod at all? I think that is out of scope of this PR, but we should capture this as an issue
> 
> Yeah, the arithmetic solver doesn't handle `mod` at all. I'll make a sub-issue.

Issue to support `mod` in arithmetic https://github.com/yaspar-org/Sundance-SMT/issues/134#issue-4731003241

File: src/arithmetic/lia/frontend.rs (hunk 2 above)
Comment (benjaminfjones) [reply src/arithmetic/lia/frontend.rs:1221]: good idea

Comment (benjaminfjones) [reply src/arithmetic/lia/frontend.rs:1221]: done!

Comment (benjaminfjones) [reply src/arithmetic/lia/frontend.rs:1221]: This is a good example because it illustrates how flooring properly changes  direction when the denominator changes sign.
